### PR TITLE
Link domain purchase to existing site flow

### LINKED ISSUE #19381: 🤖 Purchase a domain linked to existing site flow
This flow is invoked after the user chooses a new domain and then chooses to connect it to an existing site. The user should first choose the site, and then be presented with the checkout flow in a webview.
Ref: pc8HXX-1gR-p2#add-domain-choice

### PR DESCRIPTION
**Pull request by mkevins**
Fixes #19381

### Description

This PR implements the domain purchasing flow for connecting with a site. It supports two checkout scenarios: when appropriate, it navigates to the plans checkout (for free sites), otherwise it navigates to the checkout flow for a domain connected to a site with an existing paid plan.

To test:

#### Existing Paid Plan

Prerequisite: Use an account that has a site on a paid plan

1. Enter the domain management flow (`Me -> Domains`)
2. Tap the "+" button to open the domain search screen
3. Search for a domain
4. Tap a domain result to open the purchase flow
5. Choose to use the domain with an existing site
6. Pick a site which is already on a paid plan
7. Expect to be brought to the checkout screen (webview) to purchase the domain

#### Free Plan

Prerequisite: Use an account that has a site on a free plan

1. Perform steps 1-5 from above
2. Choose a site which is on a free plan
3. Expect to be brought to the plans purchase screen (webview)

#### Bonus:

Do the above tests with the sandbox store to simulate the full purchase.

## Regression Notes
1. Potential unintended areas of impact
Domain Management flow

8. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual and automated testing

9. What automated tests I added (or what prevented me from doing so)
Unit tests were added for the plans and checkout logic.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:

- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] Talkback.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] Large and small screen sizes. (Tablet and smaller phones)
- [ ] Multi-tasking: Split screen and Pop-up view. (Android 10 or higher)

### CODE REVIEW

**Comment by mkevins** [comment]
I've added the "unit-tests-exemption" label because it is problematic to unit test this contract (as it involves mocking intents, and internal behavior verification, rather than state verification).

File: WordPress/src/main/java/org/wordpress/android/ui/domains/management/purchasedomain/PurchaseDomainViewModel.kt [hunk 1]
```diff
@@ -90,27 +96,30 @@ class PurchaseDomainViewModel @AssistedInject constructor(
...
         object ErrorSubmittingCart : UiState
     }
 
     sealed class ActionEvent {
         object GoBack : ActionEvent()
         data class GoToDomainPurchasing(val domain: String) : ActionEvent()
         data class GoToSitePicker(val domain: String) : ActionEvent()
-        data class GoToExistingSite(val domain: String, val siteModel: SiteModel) : ActionEvent()
+        data class GoToExistingSiteCheckout(val domain: String, val siteModel: SiteModel) : ActionEvent()
+        data class GoToExistingSitePlans(val domain: String, val siteModel: SiteModel) : ActionEvent()
```
Comment (antonis) [inline WordPress/src/main/java/org/wordpress/android/ui/domains/management/purchasedomain/PurchaseDomainViewModel.kt:122]: Thank you for splitting and making the actions more clear with the naming 🙇

**Review by antonis — COMMENTED** [review]
Thank you for your work on this @mkevins 🙇 
The PR looks good for a draft state and I like the approach taken. I also tested the `just a domain` and the `plans` flows and was able to complete the test purchase (ref: pc8HXX-14D-p2) successfully. The the final redirection/refresh is not implemented yet the new domains were visible in the domains list when I went back 🎉 

<image src="https://github.com/wordpress-mobile/WordPress-Android/assets/304044/aaefb497-d737-4206-b59d-b90ceeef176a" width=320 />

**Comment by mkevins** [comment]
Thank you for giving this an early look Antonis! I've updated the PR description (and tests), and this is ready for another pass.

File: WordPress/src/test/java/org/wordpress/android/ui/domains/management/purchasedomain/PurchaseDomainViewModelTest.kt [hunk 2]
```diff
@@ -126,11 +129,28 @@ class PurchaseDomainViewModelTest : BaseUnitTest() {
     }
 
     @Test
-    fun `WHEN a site is chosen THEN send the GoToExistingSite action event`() = testWithActionEvents { events ->
+    @Suppress("MaxLineLength")
+    fun `WHEN a plans eligible site is chosen THEN send the GoToExistingSitePlans action event`() = testWithActionEvents { events ->
```
Comment (antonis) [inline WordPress/src/test/java/org/wordpress/android/ui/domains/management/purchasedomain/PurchaseDomainViewModelTest.kt:133]: Thank you for adding the new test cases Matthew 🙇 
The look great and cover the checkout cases logic 🥇

File: WordPress/src/test/java/org/wordpress/android/ui/domains/management/purchasedomain/PurchaseDomainViewModelTest.kt [hunk 3]
```diff
@@ -219,7 +251,7 @@ class PurchaseDomainViewModelTest : BaseUnitTest() {
         private const val cartKey = "cart_key"
         private const val siteId = 5L
         private const val supportsPrivacy = true
-        private val testSite = SiteModel().also { it.siteId = siteId }
+        private val testSite = SiteModel().apply { siteId = siteId }
```
Comment (antonis) [inline WordPress/src/test/java/org/wordpress/android/ui/domains/management/purchasedomain/PurchaseDomainViewModelTest.kt:254]: Thank for this improvement 🙇

**Review by antonis — APPROVED** [review]
Awesome work @mkevins 🙇 
I retested the flows with the  latest changes and everything worked for me. The added tests also look great 🎉

**Comment by mkevins** [comment]
Thank you for reviewing and testing Antonis!

**Pull request merged**